src/coreclr/vm/jitinterface.cpp

```diff
@@ -9323,6 +9323,532 @@ CorInfoTypeWithMod CEEInfo::getArgType (
     return result;
 }
 
+// The only difference between `CEEInfo::getArgType2` and `CEEInfo::getArgType` is
+// the arg `int *pFloatFieldFlags` which denoting whether a struct-arg using float registers.
+//
+// The value of `*pFloatFieldFlags` is returned by `CEEInfo::getFieldTypeByHnd`.
+// So the encoding details of `*pFloatFieldFlags` see `CEEInfo::getFieldTypeByHnd`.
+CorInfoTypeWithMod CEEInfo::getArgType2 (
+        CORINFO_SIG_INFO*       sig,
+        CORINFO_ARG_LIST_HANDLE args,
+        CORINFO_CLASS_HANDLE*   vcTypeRet,
+        int *pFloatFieldFlags
+        )
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    CorInfoTypeWithMod result = CorInfoTypeWithMod(CORINFO_TYPE_UNDEF);
+
+    JIT_TO_EE_TRANSITION();
+
+   _ASSERTE((BYTE*) sig->pSig <= (BYTE*) sig->args && (BYTE*) args < (BYTE*) sig->pSig + sig->cbSig);
+   _ASSERTE((BYTE*) sig->args <= (BYTE*) args);
+    INDEBUG(*vcTypeRet = CORINFO_CLASS_HANDLE((size_t)INVALID_POINTER_CC));
+
+    SigPointer ptr((unsigned __int8*) args);
+    CorElementType eType;
+    IfFailThrow(ptr.PeekElemType(&eType));
+    while (eType == ELEMENT_TYPE_PINNED)
+    {
+        result = CORINFO_TYPE_MOD_PINNED;
+        IfFailThrow(ptr.GetElemType(NULL));
+        IfFailThrow(ptr.PeekElemType(&eType));
+    }
+
+    // Now read off the "real" element type after taking any instantiations into consideration
+    SigTypeContext typeContext;
+    GetTypeContext(&sig->sigInst,&typeContext);
+
+    Module* pModule = GetModule(sig->scope);
+
+    CorElementType type = ptr.PeekElemTypeClosed(pModule, &typeContext);
+
+    TypeHandle typeHnd = TypeHandle();
+    switch (type) {
+      case ELEMENT_TYPE_VAR :
+      case ELEMENT_TYPE_MVAR :
+      case ELEMENT_TYPE_VALUETYPE :
+      case ELEMENT_TYPE_TYPEDBYREF :
+      case ELEMENT_TYPE_INTERNAL :
+      {
+            typeHnd = ptr.GetTypeHandleThrowing(pModule, &typeContext);
+            _ASSERTE(!typeHnd.IsNull());
+
+            CorElementType normType = typeHnd.GetInternalCorElementType();
+
+            // if we are looking up a value class, don't morph it to a refernece type
+            // (This can only happen in illegal IL)
+            if (!CorTypeInfo::IsObjRef(normType) || type != ELEMENT_TYPE_VALUETYPE)
+            {
+                type = normType;
+            }
+            if ((pFloatFieldFlags != NULL) && (type == ELEMENT_TYPE_VALUETYPE))
+            {
+                *pFloatFieldFlags = getFieldTypeByHnd((CORINFO_CLASS_HANDLE)typeHnd.AsTAddr());
```

Comment:
Is there a reason why the JIT cannot just call `getFieldTypeByHnd` on the value returned in `vcTypeRet` ? It would avoid all the duplication between `getArgType` and `getArgType2` methods.

Reply:
> Is there a reason why the JIT cannot just call `getFieldTypeByHnd` on the value returned in `vcTypeRet` ? It would avoid all the duplication between `getArgType` and `getArgType2` methods.

Yes, you are right.
The `getArgType2` is not the only implementation.
Now it's only called by `src/coreclr/jit/lclvars.cpp:649` within #62843 .
I can refactor it within the `src/coreclr/jit/lclvars.cpp:649` to avoid adding `getArgType2`.

Reply:
Liking the `src/coreclr/jit/lclvars.cpp:649` within #62843 ,
is there another way to access EE from the JIT.

Reply:
The JIT/EE interface is the only way to access EE from the JIT.

Reply:
> The JIT/EE interface is the only way to access EE from the JIT.

If I call the `getFieldTypeByHnd` on the value returned, I have to define `getFieldTypeByHnd` as a interface liking `getArgType2`.
It seems that there is no essential difference.

Reply:
Hi, @jkotas 
I have another question that needs your help and guidance.
How to amend the implementation of the function `getFieldTypeByHnd` to get the fields info for all cases.
Now I am amending it because I think there are some other cases to add.
Liking the cases that are not `else if (fieldType == ELEMENT_TYPE_VALUETYPE)` or `else if (fieldType == ELEMENT_TYPE_CLASS)`.

Reply:
Maybe it's better for getting your help and guidance after I update the patch.
I will update it later.

Reply:
> If I call the getFieldTypeByHnd on the value returned, I have to define getFieldTypeByHnd as a interface liking

You have it defined on that interface already: https://github.com/dotnet/runtime/pull/62893/files#diff-9219a33640f657e0bbc2f00f2aa40f6a5630999d8edd7e6d4f9bb79ff2ea8e03R2863

> It seems that there is no essential difference.

The difference is that there is no duplication between getArgType and getArgType2.

> Now I am amending it because I think there are some other cases to add.

The other cases should are primitive types. `getFieldTypeByHnd` does not return anything interesting for primitive types, so you should not need to call it for those.

Reply:
You are right.
I will amend it.

Thanks